### PR TITLE
:sparkles: (app-data): Adds data backup and restore to install / uninstall app

### PR DESCRIPTION
**Pull request by valpinkman**
### ✅ Checklist

- [x] `npx changeset` was attached.
- [ ] **Covered by automatic tests.** 
- [x] **Impact of the changes:** 
  - new commands in live-common to install and uninstall app from device with app data backup and restore
  - implement those command in LLD, transparent to the user for now (no UI implemented, just the functionality), behind a feature flag
  -  implement those command in LLM, transparent to the user for now (no UI implemented, just the functionality), behind a feature flag

### 📝 Description

This PR brings two new method in live-common for managing app install / uninstall from the device:
- `installAppWithRestore`: when installing an app, after, check if there is current app data in the provided storage and restore it
- `uninstallAppWithBackup`: when deleting an app, before, check if there is current app data and save it using the given storage

The implementation makes uses of `execWithTransport` which will be the fork where we call the new method or the old `install` / `uninstall` ones

In LLD and LLM, I implemented the new method behind the featureFlag **enableAppsBackup**. When enabled, it will transparently try to backup and restore app data if any.

**Test Conditions**
- Stax OS >= 1.5.0 | Flex OS >= 1.1.0
- Provider 83
- Boilerplate App (on device) >= 2.1.0
- Having opened the app at least once (and you can meddle with the settings then to see if it's correctly restores)
- Bonus => you can trigger **mock app update** in Developer Settings to fake updates and reinstall the current apps on your device

### ❓ Context

- **JIRA or GitHub link**: [LIVE-13453] [LIVE-12585] 

---

### 🧐 Checklist for the PR Reviewers

- **The code aligns with the requirements** described in the linked JIRA or GitHub issue.
- **The PR description clearly documents the changes** made and explains any technical trade-offs or design decisions.
- **There are no undocumented trade-offs**, technical debt, or maintainability issues.
- **The PR has been tested** thoroughly, and any potential edge cases have been considered and handled.
- **Any new dependencies** have been justified and documented.
- **Performance** considerations have been taken into account. (changes have been profiled or benchmarked if necessary)

[LIVE-13453]: https://ledgerhq.atlassian.net/browse/LIVE-13453?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by gre — DISMISSED** [review]
ok for hub scope. 

more general feedback beyond this PR, but the naming "Storage" is a bit generic and maybe could have been named something else since it only relates to device sdk 🤔

File: libs/device-core/src/commands/use-cases/app-backup/getAppStorageInfo.ts [hunk 1]
```diff
@@ -91,7 +94,7 @@ export function parseResponse(data: Buffer): AppStorageInfo {
        */
       const hasSettings = (properties & 1) === 1;
       const hasData = (properties & 2) === 2;
-      const hash = data.subarray(offset, offset + 32).toString(); // Len = 32
+      const hash = data.subarray(offset, offset + 32).toString("hex"); // Len = 32
```
Comment (jiyuzhuang) [inline libs/device-core/src/commands/use-cases/app-backup/getAppStorageInfo.ts:97]: This change needs an adjustment in the unit test.

File: libs/ledger-live-common/src/apps/runner.ts [hunk 2]
```diff
@@ -140,8 +168,16 @@ export const runAll = (state: State, exec: Exec): Observable<State> =>
     ),
     reduce(reducer, state),
   );
-export const runOneAppOp = (state: State, appOp: AppOp, exec: Exec): Observable<State> =>
-  runAppOp(state, appOp, exec).pipe(
+export const runOneAppOp = ({
```
Comment (jiyuzhuang) [inline libs/ledger-live-common/src/apps/runner.ts:171]: :+1:

File: apps/ledger-live-desktop/src/renderer/hooks/storage-provider/useStorage.tsx [hunk 3]
```diff
@@ -0,0 +1,17 @@
+import React, { createContext, useContext } from "react";
+import { DesktopStorageProvider } from "./DesktopStorageProvider";
+
+const StorageContext = createContext(new DesktopStorageProvider());
+
+export const useStorageProvider = () => {
```
Comment (jiyuzhuang) [inline apps/ledger-live-desktop/src/renderer/hooks/storage-provider/useStorage.tsx:6]: [COULD] This hook seems like not to be used twice, so putting all the initialisation in the function at line 14 can make it more clear and simpler, WDYT ?

File: apps/ledger-live-mobile/src/hooks/storageProvider/useStorage.tsx [hunk 4]
```diff
@@ -0,0 +1,17 @@
+import React from "react";
+import { MobileStorageProvider } from "./MobileStorageProvider";
+
+const StorageContext = React.createContext(new MobileStorageProvider());
+
+export const useStorage = () => {
```
Comment (jiyuzhuang) [inline apps/ledger-live-mobile/src/hooks/storageProvider/useStorage.tsx:6]: Same with [this comment](https://github.com/LedgerHQ/ledger-live/pull/7572/files#r1721954788)

File: libs/ledger-live-common/src/apps/runner.ts [hunk 5]
```diff
@@ -42,7 +50,24 @@ export const runAppOp = (
...
+        });
+      }
+
+      return exec({
+        appOp,
+        targetId: deviceInfo.targetId,
+        app,
+        modelId: deviceModel.id,
+      });
+    }).pipe(
```
Comment (jiyuzhuang) [inline libs/ledger-live-common/src/apps/runner.ts:70]: [COULD]
```suggestion
    defer(() => exec({
        appOp,
        targetId: deviceInfo.targetId,
        app,
        modelId: deviceModel.id,
        ...(storage && { storage }),
      })).pipe(
```

File: libs/ledger-live-common/src/hw/uninstallAppWithBackup.ts [hunk 6]
```diff
@@ -0,0 +1,20 @@
+import uninstallApp from "./uninstallApp";
+import { AppStorageType, StorageProvider } from "../device/use-cases/appDataBackup/types";
```
Comment (jiyuzhuang) [inline libs/ledger-live-common/src/hw/uninstallAppWithBackup.ts:2]:
```suggestion
import type { AppStorageType, StorageProvider } from "../device/use-cases/appDataBackup/types";
```

File: libs/ledger-live-common/src/hw/installAppWithRestore.ts [hunk 7]
```diff
@@ -0,0 +1,27 @@
+import installApp from "./installApp";
+import {
```
Comment (jiyuzhuang) [inline libs/ledger-live-common/src/hw/installAppWithRestore.ts:2]:
```suggestion
import type {
```

File: apps/ledger-live-desktop/src/renderer/hooks/storage-provider/useStorage.tsx (hunk 3 above)
Comment (valpinkman) [reply apps/ledger-live-desktop/src/renderer/hooks/storage-provider/useStorage.tsx:6]: I'm not sure I understand

File: libs/ledger-live-common/src/hw/uninstallAppWithBackup.ts (hunk 6 above)
Comment (valpinkman) [reply libs/ledger-live-common/src/hw/uninstallAppWithBackup.ts:2]: :+1:

File: libs/ledger-live-common/src/hw/installAppWithRestore.ts (hunk 7 above)
Comment (valpinkman) [reply libs/ledger-live-common/src/hw/installAppWithRestore.ts:2]: :+1:

File: apps/ledger-live-desktop/src/renderer/hooks/storage-provider/useStorage.tsx (hunk 3 above)
Comment (jiyuzhuang) [reply apps/ledger-live-desktop/src/renderer/hooks/storage-provider/useStorage.tsx:6]: I've made a mistake.

**Comment by jiyuzhuang** [comment]
Great job ! And thank you for fixing the `lc` parameter errors in the use cases.

**Review by jiyuzhuang — DISMISSED** [review]

**Review by mcayuelas-ledger — APPROVED** [review]
OK for Hub

**Pull request merged**